### ISSUE TITLE
Multiple tests started to fail in Firefox

### ISSUE THREAD

**Issue by OrKoN** [puppeteer/puppeteer#8361]
See https://github.com/puppeteer/puppeteer/runs/6485039988?check_suite_focus=true

it seems the failure happens consistently.

**Comment by OrKoN** [comment]
@whimboo could you please take a look? it seems it's only on Mac. I have not disabled the tests for now as there are multiple of them failing.

**Comment by whimboo** [comment]
@OrKoN is that related to all of these failing tests? We haven't landed anything for our CDP implementation which might have caused that. Since when exactly is it happening? Just with today's Nightly build or did it start to fail earlier?

**Comment by OrKoN** [comment]
I think it started ~20h ago

**Comment by OrKoN** [comment]
This CL has landed https://github.com/puppeteer/puppeteer/pull/8353 23h ago. It looks like there were some flaky failures but it passed after 3 attempts. And then once it was in main the tests started to fail consistently on Mac (I did three full run retries)

**Comment by whimboo** [comment]
I tried to run locally after updating Puppeteer (main). But I'm not able to run and install due to the following failure:

```
node:internal/errors:465
    ErrorCaptureStackTrace(err);
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:372:5)
    at validateString (node:internal/validators:120:11)
    at dirname (node:path:1276:5)
    at Object.<anonymous> (/Users/henrik/code/puppeteer/lib/cjs/puppeteer/constants.js:6:82)
    at Module._compile (node:internal/modules/cjs/loader:1099:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:999:19)
    at require (node:internal/modules/cjs/helpers:102:18) {
  code: 'ERR_INVALID_ARG_TYPE'
  ```

If you could run locally and check with the trace logging enabled I would appreciate. I clearly would need more details or the above failure fixed.

**Comment by OrKoN** [comment]
hm, which Node version are you on? I am able to reproduce the test failure locally (not this error). Btw, would it make sense to add a short documentation about how to provide debug data for Firefox (in CONTRIBUTING.md)? Also, I wonder if there is an easy way to download the nightly - 1 build to confirm that this is indeed due to a change in Firefox?

**Comment by whimboo** [comment]
I actually had to remove the `lib/` folder to get `npm install` working. It would be good if that could be done automatically when the command is run?

**Comment by OrKoN** [comment]
I have never encountered a need to delete the lib folder. I think that should not be required and I wonder what could have been wrong there? 🤔

**Comment by whimboo** [comment]
I will have a look and file a new issue with details.

Regarding the question of downloading an older Nightly build you could use [mozdownload](https://pypi.org/project/mozdownload/). Then simply run `mozdownload -t daily --date 2022-05-17` would get you a Nightly from the given date.

**Comment by whimboo** [comment]
Note that I just tried to run main latest for Puppeteer with Firefox Beta and it also fails completely for wait tasks. Was there maybe a change in Puppeteer which triggered this?

**Comment by OrKoN** [comment]
@whimboo I don't think so since every change goes through the same tests. Are those failures by any chance flaky?

**Comment by OrKoN** [comment]
I had this test failure reproduce locally in isolation:

```

  2) navigation
       Page.goBack
         should work:
     TypeError: Cannot read property 'ok' of null
      at Context.<anonymous> (test/navigation.spec.ts:637:23)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
      
  ```
  
  but it appears to be flaky and stop reproducing once I turn on DEBUG=puppeteer.*

**Comment by OrKoN** [comment]
On the complete test suite the following fails locally:

```
  1) Emulation
       Page.viewport
         should be detectable by Modernizr:
     ProtocolError: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id callFunctionOn@chrome://remote/content/cdp/domains/content/Runtime.jsm:218:15
execute@chrome://remote/content/cdp/domains/DomainCache.jsm:101:25
receiveMessage@chrome://remote/content/cdp/sessions/ContentProcessSession.jsm:84:45

      at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:304:16
      at new Promise (<anonymous>)
      at CDPSession.send (src/common/Connection.ts:300:12)
      at ExecutionContext._evaluateInternal (src/common/ExecutionContext.ts:254:44)
      at ExecutionContext.evaluate (src/common/ExecutionContext.ts:140:23)
      at DOMWorld.evaluate (src/common/DOMWorld.ts:174:20)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  2) Launcher specs
       Puppeteer
         Puppeteer.launch
           userDataDir option:
     Error: Timeout of 25000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/alexrudenko/src/puppeteer/test/launcher.spec.ts)
      at listOnTimeout (node:internal/timers:557:17)
      at processTimers (node:internal/timers:500:7)

  3) Launcher specs
       Puppeteer
         Puppeteer.launch
           userDataDir option restores preferences:
     Error: Timeout of 25000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/alexrudenko/src/puppeteer/test/launcher.spec.ts)
      at listOnTimeout (node:internal/timers:557:17)
      at processTimers (node:internal/timers:500:7)

  4) Launcher specs
       Puppeteer
         Puppeteer.launch
           userDataDir option should restore state:
     Error: Timeout of 25000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/alexrudenko/src/puppeteer/test/launcher.spec.ts)
      at listOnTimeout (node:internal/timers:557:17)
      at processTimers (node:internal/timers:500:7)

  5) Mouse
       should work with mobile viewports and cross process navigations:
     ProtocolError: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id callFunctionOn@chrome://remote/content/cdp/domains/content/Runtime.jsm:218:15
execute@chrome://remote/content/cdp/domains/DomainCache.jsm:101:25
receiveMessage@chrome://remote/content/cdp/sessions/ContentProcessSession.jsm:84:45

      at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:304:16
      at new Promise (<anonymous>)
      at CDPSession.send (src/common/Connection.ts:300:12)
      at ExecutionContext._evaluateInternal (src/common/ExecutionContext.ts:254:44)
      at ExecutionContext.evaluate (src/common/ExecutionContext.ts:140:23)
      at DOMWorld.evaluate (src/common/DOMWorld.ts:174:20)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  6) navigation
       Page.goBack
         should work:
     TypeError: Cannot read properties of null (reading 'ok')
      at Context.<anonymous> (test/navigation.spec.ts:637:23)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  7) waittask specs
       Frame.waitForFunction
         should survive cross-process navigation:
     ProtocolError: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id callFunctionOn@chrome://remote/content/cdp/domains/content/Runtime.jsm:218:15
execute@chrome://remote/content/cdp/domains/DomainCache.jsm:101:25
receiveMessage@chrome://remote/content/cdp/sessions/ContentProcessSession.jsm:84:45

      at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:304:16
      at new Promise (<anonymous>)
      at CDPSession.send (src/common/Connection.ts:300:12)
      at ExecutionContext._evaluateInternal (src/common/ExecutionContext.ts:254:44)
      at ExecutionContext.evaluate (src/common/ExecutionContext.ts:140:23)
      at DOMWorld.evaluate (src/common/DOMWorld.ts:174:20)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  8) waittask specs
       "after all" hook in "waittask specs":
     done() called multiple times in hook <waittask specs "after all" hook in "waittask specs"> of file /Users/alexrudenko/src/puppeteer/test/waittask.spec.ts; in addition, done() received error: ProtocolError: Protocol error (Runtime.callFunctionOn): Target closed.
    at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:304:16
    at new Promise (<anonymous>)
    at CDPSession.send (/Users/alexrudenko/src/puppeteer/src/common/Connection.ts:300:12)
    at ExecutionContext._evaluateInternal (/Users/alexrudenko/src/puppeteer/src/common/ExecutionContext.ts:254:44)
    at ExecutionContext.evaluateHandle (/Users/alexrudenko/src/puppeteer/src/common/ExecutionContext.ts:193:17)
    at WaitTask.rerun (/Users/alexrudenko/src/puppeteer/src/common/DOMWorld.ts:841:31)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  originalMessage: '',
  uncaught: true
}
  Error: done() called multiple times in hook <waittask specs "after all" hook in "waittask specs"> of file /Users/alexrudenko/src/puppeteer/test/waittask.spec.ts; in addition, done() received error: ProtocolError: Protocol error (Runtime.callFunctionOn): Target closed.
      at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:304:16
      at new Promise (<anonymous>)
      at CDPSession.send (src/common/Connection.ts:300:12)
      at ExecutionContext._evaluateInternal (src/common/ExecutionContext.ts:254:44)
      at ExecutionContext.evaluateHandle (src/common/ExecutionContext.ts:193:17)
      at WaitTask.rerun (src/common/DOMWorld.ts:841:31)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5) {
    originalMessage: '',
    uncaught: true
  }
      at process.emit (node:events:402:35)
      at process.emit (node:domain:475:12)
      at process.emit.sharedData.processEmitHook.installedValue [as emit] (node_modules/@cspotcode/source-map-support/source-map-support.js:613:40)
      at process._fatalException (node:internal/process/execution:164:25)
```

Also the headful tests often show a dialog that Firefox has crashed asking to send a crash report.

**Comment by whimboo** [comment]
Ok, so it looks like there was a change in Firefox which caused it to crash under certain situations. Let me try to get it reproduced (I hope that I'll be able soon). But clearly we shouldn't show the crash report dialog. Instead we should ignore and write the appropriate minidump files for investigation. How does Puppeteer handle crashes of Chrome btw?

Have you had the chance to check older Nightly builds? If yes it would be good to know when it started. If not I will have to do it once it's reproducible for me.

**Comment by OrKoN** [comment]
I have similar but fewer failures in the previous nightly:

```
 1) Emulation
       Page.viewport
         should be detectable by Modernizr:
     ProtocolError: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id callFunctionOn@chrome://remote/content/cdp/domains/content/Runtime.jsm:226:15
execute@chrome://remote/content/cdp/domains/DomainCache.jsm:101:25
receiveMessage@chrome://remote/content/cdp/sessions/ContentProcessSession.jsm:84:45

      at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:304:16
      at new Promise (<anonymous>)
      at CDPSession.send (src/common/Connection.ts:300:12)
      at ExecutionContext._evaluateInternal (src/common/ExecutionContext.ts:254:44)
      at ExecutionContext.evaluate (src/common/ExecutionContext.ts:140:23)
      at DOMWorld.evaluate (src/common/DOMWorld.ts:174:20)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  2) Mouse
       should work with mobile viewports and cross process navigations:
     ProtocolError: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id callFunctionOn@chrome://remote/content/cdp/domains/content/Runtime.jsm:226:15
execute@chrome://remote/content/cdp/domains/DomainCache.jsm:101:25
receiveMessage@chrome://remote/content/cdp/sessions/ContentProcessSession.jsm:84:45

      at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:304:16
      at new Promise (<anonymous>)
      at CDPSession.send (src/common/Connection.ts:300:12)
      at ExecutionContext._evaluateInternal (src/common/ExecutionContext.ts:254:44)
      at ExecutionContext.evaluate (src/common/ExecutionContext.ts:140:23)
      at DOMWorld.evaluate (src/common/DOMWorld.ts:174:20)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  3) waittask specs
       Frame.waitForFunction
         should survive cross-process navigation:
     ProtocolError: Protocol error (Runtime.callFunctionOn): Cannot find context with specified id callFunctionOn@chrome://remote/content/cdp/domains/content/Runtime.jsm:226:15
execute@chrome://remote/content/cdp/domains/DomainCache.jsm:101:25
receiveMessage@chrome://remote/content/cdp/sessions/ContentProcessSession.jsm:84:45

      at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:304:16
      at new Promise (<anonymous>)
      at CDPSession.send (src/common/Connection.ts:300:12)
      at ExecutionContext._evaluateInternal (src/common/ExecutionContext.ts:254:44)
      at ExecutionContext.evaluate (src/common/ExecutionContext.ts:140:23)
      at DOMWorld.evaluate (src/common/DOMWorld.ts:174:20)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  4) waittask specs
       "before each" hook for "should survive navigations":
     Uncaught ProtocolError: Protocol error (Runtime.callFunctionOn): can't access property "sendAsyncMessage", this.mm is null executeInChild/<@chrome://remote/content/cdp/sessions/TabSession.jsm:80:7
executeInChild@chrome://remote/content/cdp/sessions/TabSession.jsm:75:12
execute@chrome://remote/content/cdp/sessions/TabSession.jsm:71:17
onPacket@chrome://remote/content/cdp/CDPConnection.jsm:262:36
onMessage@chrome://remote/content/server/WebSocketTransport.jsm:89:18
handleEvent@chrome://remote/content/server/WebSocketTransport.jsm:71:14

      at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:304:16
      at new Promise (<anonymous>)
      at CDPSession.send (src/common/Connection.ts:300:12)
      at ExecutionContext._evaluateInternal (src/common/ExecutionContext.ts:254:44)
      at ExecutionContext.evaluateHandle (src/common/ExecutionContext.ts:193:17)
      at WaitTask.rerun (src/common/DOMWorld.ts:841:31)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  5) waittask specs
       "after each" hook for "should survive navigations":
     TypeError: Cannot read properties of null (reading 'close')
      at Context.<anonymous> (test/mocha-utils.ts:255:25)
      at processImmediate (node:internal/timers:464:21)

  6) waittask specs
       "before each" hook for "should survive navigations":
     done() called multiple times in hook <waittask specs "before each" hook for "should survive navigations"> of file /Users/alexrudenko/src/puppeteer/test/waittask.spec.ts; in addition, done() received error: ProtocolError: Protocol error (Target.createTarget): Target closed.
    at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:106:16
    at new Promise (<anonymous>)
    at Connection.send (/Users/alexrudenko/src/puppeteer/src/common/Connection.ts:102:12)
    at Browser._createPageInContext (/Users/alexrudenko/src/puppeteer/src/common/Browser.ts:511:49)
    at BrowserContext.newPage (/Users/alexrudenko/src/puppeteer/src/common/Browser.ts:861:26)
    at Context.<anonymous> (/Users/alexrudenko/src/puppeteer/test/mocha-utils.ts:251:38)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  originalMessage: ''
}
  Error: done() called multiple times in hook <waittask specs "before each" hook for "should survive navigations"> of file /Users/alexrudenko/src/puppeteer/test/waittask.spec.ts; in addition, done() received error: ProtocolError: Protocol error (Target.createTarget): Target closed.
      at /Users/alexrudenko/src/puppeteer/src/common/Connection.ts:106:16
      at new Promise (<anonymous>)
      at Connection.send (src/common/Connection.ts:102:12)
      at Browser._createPageInContext (src/common/Browser.ts:511:49)
      at BrowserContext.newPage (src/common/Browser.ts:861:26)
      at Context.<anonymous> (test/mocha-utils.ts:251:38)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5) {
    originalMessage: ''
  }
  ```
  
  But I wonder if the Nightly build auto-updates if installed on MacOS (mozdownload gave me a .dmg file). 
  
  >How does Puppeteer handle crashes of Chrome btw?
  
  I am actually not very sure but I think `--disable-breakpad` flag that Puppeteer sets disabled all the crash reporting. Not sure if minidumps are generated.

**Comment by whimboo** [comment]
Ok, so far no luck on my side. I'm not even able to get a connection to CDP due to the error `connect ECONNREFUSED ::1:54553`.

@OrKoN could you please run only a single of these failing tests with the following modifications?

1. Add the `'remote.log.level': 'Trace'` line to the preferences for Firefox in Launcher.spec.ts
2. Run the unit tests with `DUMPIO=1`. 

I'm interested to see the output.

**Comment by OrKoN** [comment]
Here is the output for the test `should be detectable by Modernizr`. Unfortunately, it looks like this failure is flaky https://gist.github.com/OrKoN/1d0f00e0cd1dbf420451e8d91c21f152

**Comment by whimboo** [comment]
This really looks like a crash of the target tab. @OrKoN would you mind to run Firefox with the `MOZ_CRASHREPORTER_SHUTDOWN=1` environment variable set? It should force a shutdown of Firefox under such a condition and it should be visible in the log.

**Comment by OrKoN** [comment]
the log is here https://gist.github.com/OrKoN/e22e06c802b552a8d22b49d3c373e974 it does not look like it crashes but it looks some execution contexts get destroyed earlier than the should be

**Comment by whimboo** [comment]
I'm now able to reproduce it even when it doesn't happen that often. Will try to find out now what excactly caused it.

**Comment by whimboo** [comment]
So I can see the same failure in a Firefox Beta build with latest Puppeteer. As mentioned earlier, could it be that a change in Puppeteer actually triggered this failure? I can check as well after I finished the meetings.

**Comment by OrKoN** [comment]
I don't see a good suspect PR for this failure. https://github.com/puppeteer/puppeteer/commit/7107d2d300f9a5b5c38210163ff5df42d19d8453 was the latest PR that landed before the failures started to block CI. Also, it does not seem to fail (be flaky) for Chromium 🤔

**Comment by juliandescottes** [comment]
@OrKoN testing locally this was regressed by 7107d2d300f9a5b5c38210163ff5df42d19d8453 , page.goto resolves too early for firefox and the evaluate call goes through earlier than expected. Any idea?

**Comment by OrKoN** [comment]
Hm, so then it means that Firefox dispatches either Page.frameNavigated or Page.navigatedWithinDocument before the navigation is fully finished?

**Comment by OrKoN** [comment]
From the CDP message log I see that the `Page.navigatedWithinDocument` event is sent by the browser immediately after Page.navigate if the run fails. Then also `Page.frameNavigated` is sent. In Chromium, those events are mutually exclusive so it looks like the change revealed some difference in the impl. I will continue looking into it but I wondered if you have some insights into when you dispatch those events? We had to land https://github.com/puppeteer/puppeteer/commit/7107d2d300f9a5b5c38210163ff5df42d19d8453 because of the breaking changes in Chromium where we cannot rely on the loaderId anymore. Also, relying on loaderId seemed kind of fragile.

**Comment by juliandescottes** [comment]
I think you are right, we are probably emitting `frameNavigated` too early: https://searchfox.org/mozilla-central/rev/8fe6930c0832009b3162bebee7d4ede1a4c8c9a8/remote/cdp/observers/ContextObserver.jsm#133-149

This is currently mapped to `document-element-inserted` which does not indicate a full navigation. I tried to delay the whole block, but there are still issues. I suspect we also have a race internal to our implementation between the "frame-navigated" and "context-created" event. 

I guess with the changes from https://github.com/puppeteer/puppeteer/commit/7107d2d300f9a5b5c38210163ff5df42d19d8453 we are now resolving `goto` as soon as "frame-navigated" is received, and internally we haven't had time to process "context-created". 

I don't see an easy fix for now but I'll keep investigating.

**Comment by juliandescottes** [comment]
Delaying "frame-navigated" & "context-created" is probably not the right approach. As mentioned by @OrKoN the issue comes actually from the `navigatedWithinDocument` event. So this event is emitted on window "load" https://searchfox.org/mozilla-central/rev/8fe6930c0832009b3162bebee7d4ede1a4c8c9a8/remote/cdp/domains/content/Page.jsm#366-369

This is a known issue and https://bugzilla.mozilla.org/show_bug.cgi?id=1636453 suggests to map this to hashchange instead.

**Comment by juliandescottes** [comment]
Most likely I guess a navigation resolves early because it catches a `navigatedWithinDocument` event from a previous navigation, although my failure logs don't show that clearly for now.

**Comment by OrKoN** [comment]
@juliandescottes by looking at the test I don't see any reason to send navigatedWithinDocument at all (which does look like [1636453](https://bugzilla.mozilla.org/show_bug.cgi?id=1636453)). Also, it's the second goto that seems to always fail and not the first one. Perhaps the change of the viewport triggers a `navigatedWithinDocument` event.

**Comment by juliandescottes** [comment]
Updates on my side.

I've tried to emit `navigatedWithinDocument` only on hashchange, but this makes navigations to about:blank fail with the version of puppeteer we currently vendor in our codebase (or any version earlier than 7107d2d actually). For an about:blank navigation, Firefox doesn't return a `loaderId` from Page.navigate, which means that older puppeteer versions expect a "same document navigation" and hang waiting for `navigatedWithinDocument` forever. Our loaderId is fetched from our network request interception, so I imagine all navigations which don't trigger a proper request will show a similar behavior.

One option is to try to fix all navigations in the Firefox CDP implementation to emit the proper loaderId, but that doesn't sound trivial.

So maybe for now we need to accept that Firefox emits extra `navigatedWithinDocument` events and work around that. The main problem right now is not the fact that the event itself is emitted, it's that it's emitted later than `frameNavigated` and might therefore be consumed by mistake for another navigation.

Some things we can try:
- fire `navigatedWithinDocument` earlier (in DOMContentLoaded maybe?)
- wait for a potential `navigatedWithinDocument` when we handle `frameNavigated` (either in puppeteer or Firefox)

**Comment by juliandescottes** [comment]
> @juliandescottes by looking at the test I don't see any reason to send navigatedWithinDocument at all (which does look like [1636453](https://bugzilla.mozilla.org/show_bug.cgi?id=1636453)). Also, it's the second goto that seems to always fail and not the first one. Perhaps the change of the viewport triggers a `navigatedWithinDocument` event.

I agree, the navigatedWithinDocument is sent for no reason, but sadly we rely on that for some navigations, to workaround the fact that we don't have a proper loaderId (see comment above).

And yes, I would expect the second navigation to be the one that fails, because it catches the `navigatedWithinDocument` from the previous one.

> Perhaps the change of the viewport triggers a navigatedWithinDocument event.

Not even, we really just always trigger a `navigatedWithinDocument`, for any load :(

**Comment by OrKoN** [comment]
@juliandescottes what do you say if we restore the old behaviour but only if the product is Firefox? (not sure how easy it'd be to plumb it through). Also, what do you think if we stop requiring the Firefox tests to pass on CI temporarily until we resolve this? (some other PRs are blocked by this)

**Comment by juliandescottes** [comment]
IMO, Yes disabling Firefox on CI for now sounds reasonable, as long as we fix this before the next puppeteer release.

**Comment by whimboo** [comment]
Two questions:

1. When is the next Puppeteer release targeted for? Means how much time do we have to get this fixed? As Julian pointed out this is a complex scenario which came up surprisingly now and we will have to account some time to get it fixed.
2. Would it be possible to just disable the affected tests or how wide-spread is the issue?

Thanks

**Comment by OrKoN** [comment]
@whimboo we actually wanted to release today or tomorrow but we can wait. I think the issue would not only affect tests but also the users. So as Julian pointed our we should find a workaround (see my previous comment) before the release.

**Comment by OrKoN** [comment]
Regarding the potential workaround I mentioned: it does not look like we have a precedent for having conditional logic inside src/common based on the product. I think it's not worth introducing it now. I will see if another workaround is possible.

**Comment by juliandescottes** [comment]
For what it's worth, I tried to move `navigatedWithinDocument` to DOMContentLoaded (in https://searchfox.org/mozilla-central/rev/7f729f601c0b738f870ae0ed49098f9268e250f9/remote/cdp/domains/content/Page.jsm#341-351).

So we still emit it incorrectly, but earlier. It seems to work fine with the vendored puppeteer in mozilla-central, and it also seems to work in the puppeteer CI.

It is very hacky and I'm not sure if a fix in Firefox is the good approach for this workaround, because users of the new puppeteer version might still rely on an older Firefox binary.

**Comment by OrKoN** [comment]
Let's revert https://github.com/puppeteer/puppeteer/pull/8369 then. I think the failure does not happen in M92 and it was a forward-looking fix. Let's see if we will need it for M93. It would be nice if you could look into the issue on the Firefox side as, in general, it'd be better not to rely on the loaderId to figure out which navigation happened.

**Comment by juliandescottes** [comment]
Thanks, that's great if we can revert this for now. On our side, we'll work on this in https://bugzilla.mozilla.org/show_bug.cgi?id=1636453 and we should aim to both provide the correct loader id for navigations which don't hit the network (eg about:blank) as well as moving `navigatedWithinDocument` to `hashchange`. With both, the Firefox implementation should work both with the old and new behavior from puppeteer.

**Comment by whimboo** [comment]
@OrKoN thanks for reverting these changes! We will do our best to get this fixed on our side sooner than later. Is there any definitive ETA for you when this  change needs to re-land? Thanks for letting us know.

**Comment by OrKoN** [comment]
We plan to roll M103 in a week or two. I am not sure yet if we encounter the same issue since there were some back-merges done by the backend team. It's likely though that there could be issues because of the changes to how loaderId is reported. In the case of issues at that time, there will be somewhat more pressure to release since we want to release a version for every major Chromium release. 

Let me perhaps try to roll M103 somewhat earlier and see if we have any issues. Ideally, I think it'd be nice to stop relying on loaderId to avoid the issue altogether.

**Comment by whimboo** [comment]
Sounds reasonable! If we aren't able to get our fixes landed by that time some temporary release notes may have to be added to the Puppeteer release that will mention the incompatibility of this release with Firefox. 

Nevertheless we will try our best!

**Comment by juliandescottes** [comment]
With the current patches attached to https://bugzilla.mozilla.org/show_bug.cgi?id=1636453 Firefox seems to work both with the old and the new behavior. Hopefully we can land them soon and this will unblock your work.

**Comment by juliandescottes** [comment]
@OrKoN https://bugzilla.mozilla.org/show_bug.cgi?id=1636453 should be included in the latest Nightly, so you should be able to re-land https://github.com/puppeteer/puppeteer/pull/8369

Also pushed PR #8394 to enable a few tests for firefox thanks to this change.

**Comment by OrKoN** [comment]
Thanks for implementing this so quickly!
